### PR TITLE
 made color instances immutable

### PR DESCRIPTION
**Pull request by wmira**
I took an initial stab at #56 . I probably missed some things and still cleanup on some areas are needed.

### CODE REVIEW

**Comment by Qix-** [comment]
Woo, that's a big'n. I'll take a look at it within the next few days when I'm not sick :dancer:.

Thank you for this, this was the first thing on my list.

File: index.js [hunk 1]
```diff
@@ -2,6 +2,109 @@
...
 
+var isAnyFieldsNotUndefined = function (obj, fields) {
+	return fields.reduce(function (acc, field) {
+		if (acc !== true) {
+			return obj[field] !== undefined;
+		}
+		return acc;
+	}, false);
+};
+var setValuesMap = {
```
Comment (Qix-) [inline index.js]: New Line here.

File: index.js [hunk 2]
```diff
@@ -2,6 +2,109 @@
 var convert = require('color-convert');
 var string = require('color-string');
 
+var isAnyFieldsNotUndefined = function (obj, fields) {
+	return fields.reduce(function (acc, field) {
+		if (acc !== true) {
```
Comment (Qix-) [inline index.js]: `if (!acc) {`

File: index.js [hunk 3]
```diff
@@ -2,6 +2,109 @@
...
+			return obj[field] !== undefined;
```
Comment (Qix-) [inline index.js]: `return field in obj;`

File: index.js [hunk 4]
```diff
@@ -2,6 +2,109 @@
...
+	return Object.keys(setValuesMap).reduce(function (acc, key) {
+		if (acc === undefined && isAnyFieldsNotUndefined(obj, setValuesMap[key])) {
+			return key;
+		}
+		return acc;
+	}, undefined);
+};
+
+var cloneValues = function (values) {
+	return JSON.parse(JSON.stringify(values));
```
Comment (Qix-) [inline index.js]: Not sure about this. Invoking the JSON parser has quite a bit of overhead.

File: index.js [hunk 5]
```diff
@@ -319,10 +425,27 @@ Color.prototype = {
 	},
 
 	clone: function () {
-		return new Color(this.rgb());
+		return createNewColorInstance(cloneValues(this.values));
 	}
 };
 
+Color.prototype.setChannel = function (space, index, val) {
```
Comment (Qix-) [inline index.js]: Could you explain why this method needed to be introduced? I don't feel like it's going to help much of anything.

File: test/index.js [hunk 6]
```diff
@@ -715,3 +727,14 @@ it('Exceptions', function () {
...
+describe('immutable', function () {
+	it('Never mutates old instance', function () {
+		var colorInitial = Color('#7743CE');
+		var result1 = colorInitial.darken(0.2);
+		var result2 = colorInitial.darken(0.2);
+
+		equal(colorInitial.hexString(), '#7743CE');
+		deepEqual(result1, result2);
+	});
+});
```
Comment (Qix-) [inline test/index.js]: Definitely needs more tests.

File: test/index.js [hunk 7]
```diff
@@ -715,3 +727,14 @@ it('Exceptions', function () {
 		Color({});
 	}, /Unable to parse color from object/);
 });
+
+describe('immutable', function () {
+	it('Never mutates old instance', function () {
```
Comment (Qix-) [inline test/index.js]:
```javascript
it('should never mutate old instances', function () {
```

File: test/index.js [hunk 8]
```diff
@@ -14,6 +14,18 @@ it('Color() instance', function () {
 	ok((new Color()) instanceof Color);
 });
 
+it('test', function () {
```
Comment (Qix-) [inline test/index.js]:
```javascript
it('should deep equal itself`, function() {
```

**Comment by Qix-** [comment]
Cloning the colorspace values on each object should be done manually. Yes, it introduces more code, but invoking the JSON parser and encoder for each method call is quite excessive.

As well, this should have quite a few more tests in the `immutable` section.

**Comment by wmira** [comment]
Hi Josh,

Thanks for the review. I will update the PR.

Warren

**Comment by wmira** [comment]
I have addressed all code review comments..

**Comment by Qix-** [comment]
Thanks! Landed in f385d072cb6abf379c0b95c04b9a86a94a538b2f.

**Pull request closed without merging**